paho/client.go

```diff
@@ -551,6 +596,73 @@ func (c *Client) close() {
 	c.debug.Println("acks tracker reset")
 }
 
+func (c *Client) readLoop() {
+	c.debug.Println("readLoop starting")
+
+	incoming := make(chan *packets.ControlPacket)
+	readNext := func() {
+		c.debug.Println("reading next packet")
+		packet, err := packets.ReadPacket(c.Conn)
+		if err != nil {
+			c.errors.Printf("read packet error: %v\n", err)
+			go c.error(err)
+			return
+		}
+
+		packetIdString := ""
+		if id := packet.PacketID(); id != 0 {
+			packetIdString = fmt.Sprintf(" with id=%d", id)
+		}
+		c.debug.Printf("received %s%s\n", packet.PacketType(), packetIdString)
+
+		incoming <- packet
+	}
+
+	go readNext()
+	for {
+		select {
+		case <-c.stop:
+			c.debug.Println("readLoop stopping due to stop channel")
+			return
+		case packet := <-incoming:
+			c.incomingPackets <- packet
+			go readNext()
+		}
+	}
+}
+
+func (c *Client) writeLoop() {
```

Comment:
I agree.

Reply:
I agree. I think a tricky part of this is avoiding write operations causing a panic. We could use mutexes like is being done elsewhere, but I want to think about a solution that doesn't use mutexes. I have an idea in mind for this that I might try.

Reply:
Its possible to avoid Mutexes (mostly) by being very careful about shutdown order. However this is tricky and, unless we are very careful with future PR's, its easy to create issues later. This is the main reason that I believe we need to be sure this is the right approach before adopting it.